Added Qwen2-MoE model, optimizing its performance on Gaudi

gyou2021 opened the pull request:
Added Qwen2-MoE model into optimum-habana, optimizing its performance on Gaudi with kv_cache, reuse_cache, flash attention, internal bucket, etc.

Test command:

cd optimum-habana

python -m pytest tests/test_text_generation_example.py -v -s

cd examples/text-generation/

python run_generation.py --model_name_or_path --use_hpu_graphs --use_kv_cache --reuse_cache --trim_logits --attn_softmax_bf16 --use_flash_attention --max_input_tokens 256 --max_new_tokens 256 --batch_size 2 --limit_hpu_graphs --bucket_size 128 --bucket_internal --n_iterations 2 --warmup 2 --bf16

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests?

gyou2021 commented:
@ssarkar2 I added Qwen2-MoE test code in test_text_generation_example.py according to your comments in https://github.com/huggingface/optimum-habana/pull/1162. What's more, I created a new branch qwen2-moe following the guidance of https://github.com/huggingface/optimum/blob/main/CONTRIBUTING.md. Thank you.

imangohari1 reviewed (changes requested):
Hi @gyou2021 
Thank you for this PR. 
I have gone through this and the the tests seem running fine. 
I made a change in CI tests since `Qwen1.5-7B` was already there, and instead added the Qwen2-7B. 

Please apply the patch with `git am < 0001*` and push. 
We review afterwards. 
Thanks. 
[0001-fea-ci-Added-qwen2-7b-bs-512.patch](https://github.com/user-attachments/files/16970632/0001-fea-ci-Added-qwen2-7b-bs-512.patch)

gyou2021 commented:
Hi @imangohari1 

Thank you for your comments.

I added Qwen2-7B test and removed repeated Qwen1.5-7B test in tests/test_text_generation_example.py according to 0001-fea-ci-Added-qwen2-7b-bs-512.patch. Since "error: tests/test_text_generation_example.py: does not match index" showed when I ran git am < 0001* to apply the patch, I manually modified the code and push it. 

> Hi @gyou2021 Thank you for this PR. I have gone through this and the the tests seem running fine. I made a change in CI tests since `Qwen1.5-7B` was already there, and instead added the Qwen2-7B.
> 
> Please apply the patch with `git am < 0001*` and push. We review afterwards. Thanks. [0001-fea-ci-Added-qwen2-7b-bs-512.patch](https://github.com/user-attachments/files/16970632/0001-fea-ci-Added-qwen2-7b-bs-512.patch)

imangohari1 reviewed (approved):
LGTM! 
I have tested these changes with local cmd and CI. 
they appear functional. 

This PR has also been tested via a Integration branch https://github.com/huggingface/optimum-habana/compare/main...imangohari1:optimum-habana:ig/ci_20240911?expand=1 (CI#255) and they have not introduced a regression. 

@regisss  please take a look when you had a chance.

regisss reviewed (commented):
There are quite a lot of commented code in the modeling file, can you clean that a bit please?

imangohari1 commented:
@gyou2021 Can you please address Regiss comments? Tnx

gyou2021 commented on README.md:
```diff
@@ -211,6 +211,7 @@ The following model architectures, tasks and device distributions have been vali
 | Mixtral |   | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
 | Persimmon |   | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
 | Qwen2 | <div style="text-align:left"><li>Single card</li></div> | <div style="text-align:left"><li>Single card</li></div> | <li>[language modeling](https://github.com/huggingface/optimum-habana/tree/main/examples/language-modeling)</li><li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
+| Qwen2-MoE |   | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
```
> Can you add the same in
> 
> https://github.com/huggingface/optimum-habana/blob/091c8a52655eb35a3feaea93b4b3d98b1f6551fc/docs/source/index.mdx?plain=1#L59
> 
> please?
Added. 
Thank you for your comment.

gyou2021 replied:
> Can you add the same in
> 
> https://github.com/huggingface/optimum-habana/blob/091c8a52655eb35a3feaea93b4b3d98b1f6551fc/docs/source/index.mdx?plain=1#L59
> 
> please?

Added.
Thank you for your comment.

gyou2021 commented:
> There are quite a lot of commented code in the modeling file, can you clean that a bit please?
Thank you for your comment.
The code has been cleaned.

gyou2021 commented:
> @gyou2021 Can you please address Regiss comments? Tnx

Sure. Done.

regisss commented:
@gyou2021 There are merge conflicts to solve

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_1316). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

gyou2021 commented:
> @gyou2021 There are merge conflicts to solve edit: solved

Thank you.
